AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+    { storage.size() } -> SameAs<size_t>;
+    { storage[index] } -> ConvertibleTo<WordType&>;
+    { storage.data() } -> ConvertibleTo<WordType*>;
+};
+
+template<typename T>
+concept IntegerReadonlyStorage = IntegerStorage<T, NativeWord const>;
+// clang-format on
+
+template<size_t total_size>
+struct StackBuffer {
+    NativeWord data[total_size];
+    NativeWord* free_ptr = data;
+
+    constexpr NativeWord* allocate(size_t size)
+    {
+        NativeWord* result = free_ptr;
+        free_ptr += size;
+        VERIFY(free_ptr <= data + total_size);
+        return result;
+    }
+};
+
+template<>
+struct StackBuffer<0> {
+    NativeWord* allocate(size_t) { VERIFY_NOT_REACHED(); }
+};
+
+inline StackBuffer<0> g_null_stack_buffer;
+
+template<typename Word, bool is_signed_>
+struct StorageSpan : Span<Word> {
+    using Span<Word>::Span;
+
+    constexpr static bool is_signed = is_signed_;
+
+    explicit constexpr StorageSpan(Span<Word> span)
+        : Span<Word>(span)
+    {
+    }
+
+    constexpr bool is_negative() const
+    {
+        return is_signed && this->last() >> (word_size - 1);
+    }
+};
+
+using UnsignedStorageSpan = StorageSpan<NativeWord, false>;
+using UnsignedStorageReadonlySpan = StorageSpan<NativeWord const, false>;
+
+namespace Detail {
+// Sometimes we want to know the exact maximum amount of the bits required to represent the number.
+// However, the bit size only acts as a hint for wide multiply operations. For all other purposes,
+// `bit_size`-sized and `ceil(bit_size / word_size) * word_size`-sized `StaticStorage`s will act the
+// same.
+template<bool is_signed_, size_t bit_size>
+requires(bit_size <= max_big_int_length * word_size) struct StaticStorage {
+    constexpr static size_t static_size = (bit_size + word_size - 1) / word_size;
+    constexpr static bool is_signed = is_signed_;
+
+    // We store integers in little-endian regardless of the host endianness. We use two's complement
+    // representation of negative numbers and do not bother at all if `bit_size % word_size != 0`
+    // (i. e. do not properly handle overflows).
+    NativeWord m_data[static_size];
+
+    constexpr bool is_negative() const
+    {
+        return is_signed_ && m_data[static_size - 1] >> (word_size - 1);
+    }
+
+    constexpr static size_t size()
+    {
+        return static_size;
+    }
+
+    constexpr NativeWord operator[](size_t i) const
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord& operator[](size_t i)
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord const* data() const
+    {
+        return m_data;
+    }
+
+    constexpr NativeWord* data()
+    {
+        return m_data;
+    }
+};
+
+// There is no reason to ban u128{0} + 1 (although the second argument type is signed, the value is
+// known at the compile time to be non-negative). In order to do so, we provide overloads in
+// UFixedBigInt which take IntegerWrapper as an argument.
```

Comment:
this comment is in the wrong place, it should be over the usage of compiletime_fail, not the declaration.

Also, this declaration is essentially the same as the one in https://github.com/SerenityOS/serenity/blob/master/AK/CheckedFormatString.h#L44

We should move the one in CheckedFormatString.h to StdlibExtras or StdLibExtrasDetails in a prior commit.